### DISCUSSION TITLE
Little annoyance that "host" is missing

### DISCUSSION THREAD

**Discussion by fraternl** [Freetz-NG/freetz-ng#891] [Unerfüllte Wünsche]
I work on many different linux machines and my "muscle memory" starts typing "_**host**_" whenever I want to find out the IP of a certain hostname or the IP to get its reverse DNS.

I know it is missing on the Freetz distribution and occasionally I check if I can install it.
I also know the command is deprecated and I should really use another tool, but so is **ifconfig** (I should us **ip**) and other tools which I regularly use in bash scripts.

AFAIK it's not part of busybox and it's supposed to be a supporting tool of **bind**.
So is **dig**
The **bind** package can be installed, but I will also have named, which I don't need.
Anyhow it seems it doesn't include **host** anyway,
I could install it to get the tool **dig**

_Can **host** be added to packages?_

Do you know of another way beside a wrapper around **nslookup** to get rid of the annoyance of not having **host**?
I can write a simple wrapper in bash, I think.

Is it hard to get the "**dig**" binary as a separate tool?

**Comment by fraternl** [comment]
BTW....

I was able to get hold of the binary **dig** in a quick and dirty way
I compiled the **bind** package and copied the **dig** binary to a ~/freetz-ng/make/pkgs/package/files/root/usr/bin folder.

Then I compiled the firmware without the  bind package and installed it on a fritzbox.
Luckily the dig package didn't need any external library.

Alas, the **bind** package doesn't include **host**

```
root@fritz:/var/mod/root# dig google.nl

; <<>> DiG 9.16.44 <<>> google.nl
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 48759
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.nl.                     IN      A

;; ANSWER SECTION:
google.nl.              300     IN      A       142.250.179.195

;; Query time: 40 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Sat Nov 18 12:50:52 CET 2023
;; MSG SIZE  rcvd: 54

```

I would really like to have this more elegantly.
I don't know how to actually do it, but a new package could be created by making a copy of the bind package and then only using the dig binary.

The host.c package is still part of the used bind package.

Could someone help me integrate host.c from the bind package as standalone tool in an elegant way??

**Comment by fraternl** [comment]
Without really knowing what I'm doing I patched the files 

freetz-ng/make/pkgs/bind/external.files
freetz-ng/make/pkgs/bind/external.in
freetz-ng/make/pkgs/bind/Config.in

cat ~/freetz-ng/make/pkgs/bind/external.files
```
[ "$EXTERNAL_FREETZ_PACKAGE_BIND_NAMED"     == "y" ] && EXTERNAL_FILES+=" /usr/sbin/named"
[ "$EXTERNAL_FREETZ_PACKAGE_BIND_RNDC"      == "y" ] && EXTERNAL_FILES+=" /usr/sbin/rndc"
[ "$EXTERNAL_FREETZ_PACKAGE_BIND_NSUPDATE"  == "y" ] && EXTERNAL_FILES+=" /usr/bin/nsupdate"
[ "$EXTERNAL_FREETZ_PACKAGE_BIND_DIG"       == "y" ] && EXTERNAL_FILES+=" /usr/bin/dig"
[ "$EXTERNAL_FREETZ_PACKAGE_BIND_HOST"      == "y" ] && EXTERNAL_FILES+=" /usr/bin/host"

```

diff make/pkgs/bind/external.in ../freetz-ng-fresh/make/pkgs/bind/external.in
```
11d10
<                /usr/bin/host
46,53d44
<
<       config EXTERNAL_FREETZ_PACKAGE_BIND_HOST
<               depends on FREETZ_PACKAGE_BIND_HOST
<               bool "host"
<               default n
<               help
<                       externals the following file(s):
<                        /usr/bin/host
```



grep -A999 FREETZ_PACKAGE_BIND_DIG ~/freetz-ng/make/pkgs/bind/Config.in
```
        config FREETZ_PACKAGE_BIND_DIG
                bool "dig (binary only)"
                default n
                help
                        Adds the dig DNS lookup utility.

        config FREETZ_PACKAGE_BIND_HOST
                bool "host (binary only)"
                default n
                help
                        Adds the host DNS lookup utility.

endif # FREETZ_PACKAGE_BIND

```

With make menuconfig I then got this

![image](https://github.com/Freetz-NG/freetz-ng/assets/8692856/05b12b02-2791-44bf-96bd-b84052a9ba58)

I did a compile and I received those 2 binaries **host** and **dig**

Again I copied them to ~/freetz-ng/make/pkgs/package/files/root/usr/bin
And did a new compile without the bind package.

Now I have both "**dig**" and "**host**" on my fritzboxes....

I would really like to have those without this workaround of installing bind.
Furthermore it's more future proof and others can enjoy it too.


host isn't working properly, though
It resolves the host name but then throws an error.

```
root@fritz:/usr/bin# host google.nl
google.nl has address 142.250.179.195
socket.c:3154: REQUIRE(readable || writeable) failed.
Aborted

```

**Comment by fraternl** [comment]
I can use **nslookup** on a Fritz!box.
I thought I would be able to get a decent work "**host**" package on the Fritz!box, but I lack the knowledge to get it fixed.
This far I came and it's good enough for me.

I decided to write a small wrapper which gets rid of the runtime error message.
I renamed the host binary to **/usr/bin/hhost**
and created this small executable shell script named host

cat **/usr/bin/host**
```
#!/bin/sh
ERROR=`mktemp`
/usr/bin/hhost $* 2>${ERROR}
sed -i '/Aborted/d;/socket.c/d' ${ERROR}
if [ -s "${ERROR}" ] ; then
  cat ${ERROR} >&2
  rm  ${ERROR}
  return 1
fi
rm ${ERROR}
```

I also preserved the error state in case it's something else than this runtime error.
If someone wants to make an effort to fix this properly, he's more than welcome.

Even if it's merely to get those binaries **dig** and **host** without the kitchen sink (_named_).

**Comment by fda77** [comment]
I dont understand the problem, im using "dig" since a long time but unbound as a resolver

**Comment by fraternl** [comment]
I've explained this.
It's not about what's the best program...
I'm used to typing '**host**' and it works on all distros, except these fritzboxes.

what always happens is that I type "**host**" and then see it doesn't work and then I type **nslookup** which works.


Furthermore **dig** is only selectable if I check "**bind**".
I don't want bind on my box. It takes up unnecessary space.

Maybe you can only make it possible to have **host** and **dig** without **bind**?

**Comment by fda77** [comment]
> Furthermore dig is only selectable if I check "bind".

Why dont you do it like i do since a long time?

Seems this is for the wishlist

**Comment by fda77** [comment]
When you added a command "x" and this command does not what you expected, i see 1 options:
 - you did it wrong, check it and do it the right way
 - the program is brokren, report it upstream 

but first you should bumpo it to the latest version: https://freetz-ng.github.io/freetz-ng/SUPPORT.html#package-bump

**Comment by fraternl** [comment]
It's all there in this thread if you read it. I wrote it while I was working on it.
I now have a solution which works for me. It's not elegant, but it works.
Nothing needs to be done if you think others aren't interested in it.

I would think at least some people are interested in **dig** without the named package.

**host** was never part of Freetz.
It is however in every distro (full-blown ones) I know of and I'm used to it.
I often type it in a random linux terminal window I have open.
If I chose a window which has a terminal open to Freetz it will throw an error because **host** is not part of that distro.
That's why I want to make it part of it.

**host** is part of the **bind** package and so is **dig**

**dig** can be had by selecting the **bind** package.
**host** can not be had with Freetz.

I changed some lines so I can get **host**.  It turns out it throws a runtime error I can't solve.
I want it anyhow, so I wrote a small wrapper which gets rid of that error message.

**Comment by fda77** [comment]
Bump to the latest version and check if the problem with host still exists.

**Comment by fraternl** [comment]
I gave up with patching the bind package and trying to get '**host**' to work on the FB.
Beside the error message it throws it will significantly increase the used space and that's just not worth it.

I wrote a wrapper in bash around '**nslookup**' which should give a valid response if you feed it the same info as '**host**'

Don't treat it as an official host package.
It's only intended to give a response if you casually type in host on a Freetz terminal.
I may be the only one who has this specific annoyance.


nslookup has to accept options, so enable NSLOOKUP_BIG
```
FREETZ_BUSYBOX_NSLOOKUP=y
FREETZ_BUSYBOX___V136_NSLOOKUP=y
FREETZ_BUSYBOX___V136_FEATURE_NSLOOKUP_BIG=y
```


cat /usr/bin/host
```
#!/bin/sh
while getopts t: OPTION ; do
  case $OPTION in
  t)   Q="-q=$OPTARG";;
  ?)   printf "Usage: %s: $0 [-t type] <host> [<nameserver>]\n"
       exit 2;;
  esac
done
shift $(($OPTIND - 1))

TMPDIR=`mktemp -t -d ${0//*\/}.XXXXXXXXXX`

if [ -z "${Q}" ] ; then
  # workaround for bad behaviour when no question is given
  # I'm then asking both A, AAAA and mx records seperately.
  nslookup -q=a $1 $2 2>${TMPDIR}/error | sed 's/\t*canonical name =/ is an alias for/g;s/:53$/#53/g' >${TMPDIR}/standard
  ERRORLEVEL=$?
  nslookup -q=aaaa $1 $2 | tail -n+4 | egrep -v '(Non-authoritative|canonical)' >>${TMPDIR}/standard
  nslookup -q=mx $1 $2 | grep exchanger >>${TMPDIR}/standard
else
  nslookup $Q $1 $2 2>${TMPDIR}/error | sed 's/\t*canonical name =/ is an alias for/g;s/:53$/#53/g' >${TMPDIR}/standard
  ERRORLEVEL=$?
fi

touch ${TMPDIR}/host
if [ $2 ] && grep -q '^Server' ${TMPDIR}/standard ; then
  echo "Using domain server:" >>${TMPDIR}/host
  head -n3 ${TMPDIR}/standard | sed 's/Server:\t\t/Name: /g;s/\t/ /g' >>${TMPDIR}/host
fi
tail -n+4 ${TMPDIR}/standard | egrep -iv '(authorit|^$)' >${TMPDIR}/payload

if grep -q 'SERVFAIL' ${TMPDIR}/payload ; then
  echo "Host $1 not found: 2(SERVFAIL)" >${TMPDIR}/host
elif grep -q 'NXDOMAIN' ${TMPDIR}/payload ; then
  echo "Host $1 not found: 3(NXDOMAIN)" >${TMPDIR}/host
elif grep -q '^Address' ${TMPDIR}/payload ; then
  ENDRESULT=$1
  if grep 'is an alias for' ${TMPDIR}/standard >>${TMPDIR}/host ; then
    ENDRESULT="`grep -o 'alias for.*' ${TMPDIR}/host | awk '{print $3}'`"
  fi
  grep '^Address' ${TMPDIR}/payload  | cut -b10- >${TMPDIR}/arecords
  while read ADDRESS ; do
    if echo ${ADDRESS} | grep -q '[0-9a-f][0-9a-f][0-9a-f][0-9a-f]:[0-9a-f]' ; then
      ATYPE='has IPv6 address'
    else
      ATYPE='has address'
    fi
    echo "$ENDRESULT ${ATYPE} `echo ${ADDRESS}`" >>${TMPDIR}/host
  done<${TMPDIR}/arecords
  grep 'exchanger' ${TMPDIR}/payload >${TMPDIR}/exchangers
  # in case host is called without options
  while read LINE ; do
    echo ${LINE} | sed 's/[ \t]*mail exchanger =/ mail is handled by/g' >>${TMPDIR}/host
  done<${TMPDIR}/exchangers
else
  while read LINE ; do
    echo ${LINE} | sed 's/\t*text =/ descriptive text/g;s/\t*canonical name =/ is an alias for/g;s/\t*nameserver =/ name server/g;s/[\t ]*mail exchanger =/ mail is handled by/g' >>${TMPDIR}/host
  done<${TMPDIR}/payload
fi

[ -s ${TMPDIR}/host ] && cat ${TMPDIR}/host >&1
[ -s ${TMPDIR}/error ] && cat ${TMPDIR}/error >&2

rm -r ${TMPDIR}
exit ${ERRORLEVEL}
```

**Comment by fraternl** [comment]
I found out that the busybox nslookup doesn't behave properly anymore after enabling "BIG" with FREETZ_BUSYBOX___V136_FEATURE_NSLOOKUP_BIG=y

The vanilla nslookup doesn't accept options like q=txt, but is solid with giving a usable answer.
However, to query specific records like soa, txt, cname, aaaa, a or mx you need the "BIG" version.
The SMALL version also doesn't properly handle AAAA-records

When you call nslookup without querying a specific type of record, it doesn't give me an IPv6 address and it throws an error

```
nslookup yahoo.com
Server:         127.0.0.1
Address:        127.0.0.1:53

Non-authoritative answer:
Name:   yahoo.com
Address: 74.6.231.21
Name:   yahoo.com
Address: 74.6.231.20
Name:   yahoo.com
Address: 74.6.143.25
Name:   yahoo.com
Address: 98.137.11.164
Name:   yahoo.com
Address: 74.6.143.26
Name:   yahoo.com
Address: 98.137.11.163

*** Can't find yahoo.com: No answer

```

Without the "BIG" option, the nslookup command is pretty useless.
In my wrapper (above) I solved it by calling nslookup thrice (-q=a, q=aaaa and q=mx) if no specific query is done.
It also handles other quirks of the busybox nslookup

host yahoo.com
```
yahoo.com has address 74.6.231.21
yahoo.com has address 74.6.231.20
yahoo.com has address 74.6.143.25
yahoo.com has address 98.137.11.164
yahoo.com has address 74.6.143.26
yahoo.com has address 98.137.11.163
yahoo.com has IPv6 address 2001:4998:44:3507::8001
yahoo.com has IPv6 address 2001:4998:24:120d::1:1
yahoo.com has IPv6 address 2001:4998:44:3507::8000
yahoo.com has IPv6 address 2001:4998:124:1507::f001
yahoo.com has IPv6 address 2001:4998:24:120d::1:0
yahoo.com has IPv6 address 2001:4998:124:1507::f000
yahoo.com  mail is handled by 1 mta7.am0.yahoodns.net
yahoo.com  mail is handled by 1 mta6.am0.yahoodns.net
yahoo.com  mail is handled by 1 mta5.am0.yahoodns.net

```

**Comment by fda77** [comment]
I dont use this Busybox applet. If it is buggy ask Busybox 
But i should be much simpler to use "host" of "bind", for me it works

**Comment by fraternl** [comment]
What do you mean "_for me it works_"?
I'm talking about "host" on a freetzed box.
It can't even be selected in "make menuconfig" and after I patched Config.in, external.in, external.files I was able to get a binary for each kernel version, but it didn't copy it to the build.

Still, that version threw an error.  I have no clue why it doesn't throw that same error on yours. I did tests with a complete vanilla .config and a fresh clone of freetz-ng master yesterday. 

Would you care to explain how you got "**host**" on your box. Maybe you are using a different package?

The "_buginess_" of nslookup is well known for some years. I do not have hope it will be fixed. It seems it's used in some docker and they have the same problem.

Anyhow...  In my wrapper I wrote around it and it now behaves exactly the same as the original host package.
It also uses less resources as the version in the bind package loads several libraries.

> But i should be much simpler to use "host" of "bind", for me it works

Not after the script is written...
I now only have to enable the "BIG" options for nslookup.

**Comment by fda77** [comment]
I've added it and it works. I dont know what you bug is, maybe ive not searched enough

**Comment by fraternl** [comment]
> I've added it and it works.

I have no idea how that's possible.
There haven't been any changes on github regarding bind.
After one changes the necessary files in make/pkgs/bind/   the binary package host compiles.

It then comes with this runtime error on the fritzbox.
```
root@fritz:/usr/bin# host google.nl
google.nl has address 142.250.179.195
socket.c:3154: REQUIRE(readable || writeable) failed.
Aborted
```

**Comment by fda77** [comment]
ive no idea what youve done

**Comment by fraternl** [comment]
```
cd ~
mv freetz-ng freetz-ng.bkp
git clone https://github.com/Freetz-NG/freetz-ng.git
cd freetz-ng
make menuconfig
```

There is no **host** when I check the bind package.

**Comment by fda77** [comment]
Sure, you did not create a push-request. Or missed i something?

**Comment by fraternl** [comment]
> I dont use this Busybox applet. If it is buggy ask Busybox

Busybox abandoned Bugzilla, I think, so now I have applied to the busybox' mailing list.

I have no idea what the busybox nslookup wants to do after resolving the A-records. I've never seen any output of nslookup after giving the records.
It's not a problem, really. I've been using nslookup only on Fritzboxes, because it doesn't have "host".
I now have host and it's (probably) using less resources.
I hardly ever need the extra info that I can get with "dig"

I'm using "host" at least 10 times a day.

**Comment by fraternl** [comment]
I noticed you modified the bind package 2 days ago.
It is now possible to get the host package without patching the files within the bind package.

I gave it another try, but I'm still getting the same runtime error.
Maybe it is dependent on something that isn't properly described which is present in your firmware and not in mine?

```
host yahoo.com
yahoo.com has address 98.137.11.163
yahoo.com has address 74.6.143.26
yahoo.com has address 74.6.231.20
yahoo.com has address 74.6.143.25
yahoo.com has address 74.6.231.21
yahoo.com has address 98.137.11.164
socket.c:3154: REQUIRE(readable || writeable) failed.
Aborted

```
Because I already have a solution with my wrapper around busybox' nslookup and because a firmware with that wrapper uses much less memory (4 MB difference), I am sticking with the wrapper. 

I honestly appreciate the effort you have put into it and I do understand you want to solve it without such a wrapper.
I also understand that this whole "issue" is not an issue at all for most. Most people probably don't care at all that there's no host package.

```
-rw-r--r--  1 freetz freetz  43M Dec  3 08:55 7490_07.57.all_freetz-ng-22941M-6577f98d4_20231203-085426.image
-rw-r--r--  1 freetz freetz  39M Dec  3 09:05 7490_07.57.all_freetz-ng-22941M-6577f98d4_20231203-090432.image

```

The one with the bind package is 43 MB and the one with the wrapper script is 39 MB.

The wrapper:
```
# host yahoo.com
yahoo.com has address 98.137.11.163
yahoo.com has address 74.6.231.21
yahoo.com has address 74.6.231.20
yahoo.com has address 74.6.143.25
yahoo.com has address 74.6.143.26
yahoo.com has address 98.137.11.164
yahoo.com has IPv6 address 2001:4998:24:120d::1:0
yahoo.com has IPv6 address 2001:4998:124:1507::f001
yahoo.com has IPv6 address 2001:4998:44:3507::8000
yahoo.com has IPv6 address 2001:4998:44:3507::8001
yahoo.com has IPv6 address 2001:4998:124:1507::f000
yahoo.com has IPv6 address 2001:4998:24:120d::1:1
yahoo.com mail is handled by 1 mta6.am0.yahoodns.net
yahoo.com mail is handled by 1 mta5.am0.yahoodns.net
yahoo.com mail is handled by 1 mta7.am0.yahoodns.net

```
Thanks!
